Add 64-bit support

steveatinfincia opened the pull request:
This adds support for 64-bit versions of Java and the various installed browsers we attempt to find and use on the system. In addition, support for Edge has been added but not enabled yet.

64-bit support is handled by checking if the *operating system* is 64-bit or 32-bit rather than the FreenetTray process, and by using explicit 32-bit and 64-bit "views" of the registry that will always return the correct results we need.

Doing this required updating the .NET target to 4.0, which is the highest version supported on Windows XP but also the first version to include APIs for executing registry queries on 32/64-bit views. 

At one point it was likely that many users would have .NET 3.5 preinstalled, and therefore would not need to run a .NET installer before using Freenet, but this was only ever true on Windows 7/Server 2008 R2, and even there it was an optional component that could be removed. It is not true of Windows 8/10 which only come with 4.5/4.6/4.7 preinstalled. 

Because all .NET 4.0 targeted assemblies work as-is on any 4.x framework, as time goes by targeting 4.0 should automatically work for newer versions of Windows, while targeting 3.5 will actually require *more* users to run a .NET installer before they can use Freenet. 

However in all cases, Microsoft has stressed that [Windows is not a .NET Framework delivery channel](https://blogs.msdn.microsoft.com/oldnewthing/20110404-00/?p=11043/), so we can't rely on *any* version being preinstalled.

We were already bundling a .NET 3.5 installer with our Innosetup Freenet installer prior to this, so I have updated it to 4.0 in the changes I've made there.

steveatinfincia committed:
Update required .NET runtime to 4.0

This is the last version that is compatible with Windows XP, but also
includes some critical APIs like the ability to accurately determine
whether the operating system itself is 64-bit, as well as the ability to
query the 32-bit and 64-bit registry views separately and directly,
regardless of whether the current process is 32-bit or 64-bit

steveatinfincia committed:
Update NuGet packages to newest .NET 4.0 versions

steveatinfincia committed:
Allow VS to update the version listed in designer files

steveatinfincia committed:
Add logging class to facilitate debug logging separate from NLog

NLog is supposed to support VS console debugging, but it frequently
doesn't work properly

steveatinfincia committed:
Add a common base class for browsers

steveatinfincia committed:
Initialize FNLog at launch

steveatinfincia committed:
Add class for finding the best available JRE

Attempts to find a 64-bit JRE if the OS is 64-bit, falls back to 32-bit
if a 32-bit JRE is available

Should work regardless of whether the tray process is 32-bit or 64-bit

steveatinfincia committed:
Determine which wrapper binary to use based on the available JREs

Note: this code assumes a 64-bit wrapper binary is available, which is
true on new installs but may not be true on all installations until
updates can be pushed out through fred

steveatinfincia committed:
Add MissingJRE to list of thrown exceptions in comment

steveatinfincia committed:
Use FNLog in NodeController

steveatinfincia committed:
Use FNLog in BrowserUtil

steveatinfincia committed:
Use FNLog in CommandsMenu

steveatinfincia committed:
Add browser class for Edge

Unused for now as it needs some changes before incognito will work, and
has some quirks related to the lack of a dedicated .exe for the browser.

A URL scheme is used with explorer.exe to open it, but this doesn't
always work, for example when opening Fproxy with a URL query

steveatinfincia committed:
Convert InternetExplorer to a Browser subclass

steveatinfincia committed:
Convert Chrome to a Browser subclass

steveatinfincia committed:
Add support for finding 64-bit Opera browser

steveatinfincia committed:
Convert Opera to a Browser subclass

steveatinfincia committed:
Convert Firefox registry search code to use explicit hives & views

steveatinfincia committed:
Add Firefox fallback search paths

steveatinfincia committed:
Convert Firefox to Browser subclass, use registry views and add fallback

steveatinfincia committed:
Add appveyor configuration

steveatinfincia committed:
Always set/delete startup arguments in the 32-bit registry view

The installer is a 32-bit process, so by default it sets the "run at
login" registry key in the 32-bit registry view.

On 64-bit Windows, the tray app won't be able to delete that key without
explicitly checking the 32-bit registry view.

steveatinfincia committed:
Set autorun config to match what the installer was already doing

This prevents the preferences checkboxes from being out of sync with the
state of the registry and preferences on a new installation

steveatinfincia committed:
Migrate preferences from older assembly versions at launch

Without this, any preferences set by a user will be reset to the default
values every time the assembly version changes, for example upgrading
the tray app would cause it to forget user preferences

steveatinfincia committed:
v1.1.0.0

Thynix commented:
I've done a preliminary review of this, but want to do another pass still. The only thing that stood out to me as odd is the FNLog class, but it's not weird enough to block and it'll be fine to get rid of it once NLog fixes their bug. When I was working with it I just reopened the log file, but I can understand wanting to actually fix the problem. :)

steveatinfincia committed:
Remove reference to System.Deployment

This assembly isn’t used anywhere and prevents building the Windows
releases on Linux using Mono

steveatinfincia committed:
Add Linux/Windows build instructions to readme

steveatinfincia committed:
v1.1.1.0

steveatinfincia commented:
@Thynix sorry for not replying to this until now, Arne pointed out your comment here yesterday but I never got a notification at the time :\ 

I don't think I have access to the Coverity page you linked, if it's not easy to make it accessible just let me know the line numbers in the code and I can fix them.

I agree about FNLog too, that can go away soon.

Thynix commented:
No worries. I've approved your Coverity access request.

steveatinfincia committed:
Fix inverted null check reported by Coverity

CID 171805

steveatinfincia committed:
Check native registry view for Windows version

This fixes Coverity CID 171806

On Windows the CurrentVersion key is only present in the native view
of the registry, so a 32-bit application checking the
windows version on 64-bit Windows needs to check the 64-bit registry
view explicitly.

In practice, the wintray app will never be running in 32-bit mode on
64-bit Windows because we let it run in AnyCPU mode.

The default behavior of checking the native registry view should have
been correct as a result, but it is technically possible for the app
to be run in x86 mode which is where the Coverity defect comes from.

This will make sure the code does the right thing either way.

steveatinfincia committed:
Fix static analysis warning

Property getters shouldn’t throw exceptions

steveatinfincia commented:
@Thynix Cool :)

I've fixed both of the Coverity defects, one was intentional due to the way we run the process, but I changed the code so it won't be an issue either way.

steveatinfincia committed:
v1.2.0.0
